Add WebSocket.Factory option to Connection.Options

chrisfillmore opened the pull request:
Closes #21

chrisfillmore committed:
Add WebSocket.Factory option to Connection.Options

Closes #21

GediminasZukas commented on library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt:
```diff
@@ -38,7 +39,8 @@ class Connection constructor(
             var reconnectionDelay: Int = 3,
             var reconnectionDelayMax: Int = 30,
-            var okHttpClientFactory: OkHttpClientFactory? = null
+            var okHttpClientFactory: OkHttpClientFactory? = null,
+            var okHttpWebSocketFactory: WebSocket.Factory? = null
```
Hi, thanks for the PR!

Would it be possible to have single option field for providing `WebSocket.Factory`? As current solution would allow to provide both `okHttpClientFactory` and `okHttpWebSocketFactory` and silently `okHttpWebSocketFactory` is preferred. I think this would be confusing library API. 

I see we set `sslContext` and `hostnameVerifier` options on `okHttpClientFactory` but maybe we could force to do it on the client side (instead of doing it internally) if custom factory is provided to the library.

chrisfillmore replied:
Hi @GediminasZukas, thanks for the review!

Just so I understand correctly, you would prefer if the client could configure just one field for `OkHttpClientFactory` and `WebSocket.Factory`? Please correct me if I've misunderstood.

It's not possible to provide a custom `WebSocket.Factory` by configuring `OkHttpClient`. You also cannot extend `OkHttpClient`, since actioncable-client-kotlin will call `newBuilder()`:

https://github.com/vinted/actioncable-client-kotlin/blob/c8765676818d9e4d73e3ea26f6664df6e464fccf/library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt#L124-L125

This overrides any customization the client made (calling `build()` will return `new OkHttpClient()`. Also, clients are discouraged from extending `OkHttpClient`, I checked with OkHttp here: https://github.com/square/okhttp/issues/6242

I believe the only solution is to provide a separate field. I had the same PR with Ktor, here: https://github.com/ktorio/ktor/pull/2035

Please let me know if you have some suggestion. Thanks.

GediminasZukas replied:
> Just so I understand correctly, you would prefer if the client could configure just one field for OkHttpClientFactory and WebSocket.Factory? Please correct me if I've misunderstood.

Yes, you are basically correct. Actually I am thinking that maybe would be possible to move full building of `WebSocket.Factory` implementation to the client side instead of doing it partially internally. Here is the idea how it could look like:

**Library:**

single option field for factory: `var websocketFactory: WebSocket.Factory`

```
val websocketFactory = (options.websocketFactory?.invoke() 
         ?: OkHttpClient()).newBuilder().sslSocketFactory(it.socketFactory).hostnameVerifier(it).build() // no custom factory provided, build client internally with supplied option params
```

**Client:**

```
options.connection.websocketFactory = {
    OkHttpClient().newBuilder()
                    .addNetworkInterceptor(StethoInterceptor()) 
                    .sslSocketFactory(it.socketFactory) // mention in docs to set directly on custom client
                    .hostnameVerifier(it) // mention in docs to set directly on custom client
                    .build()
}
```

or

```
options.connection.websocketFactory = {
    object : WebSocket.Factory {
        // your custom implementation
     }
}
```

Would such implementation solve the problem?

chrisfillmore replied:
If I understand correctly, what you propose would be to change the connection options, so that the current field:

```
// OkHttpClientFactory is typealias for () -> OkHttpClient
var okHttpClientFactory: OkHttpClientFactory?
```
would become a new field:
```
var webSocketFactory: () -> WebSocket.Factory
```
And we would remove the old `okHttpClientFactory` field? And we would update docs to tell clients they need to configure OkHttpClient in their code.

This is fine for me, but would be a breaking change for other users. Is that ok with you?

GediminasZukas replied:
Yes, you are correct. I think we can drop that typealias too. Also lambda type seems not needed, probably could be just: `var webSocketFactory: WebSocket.Factory`

> And we would update docs to tell clients they need to configure OkHttpClient in their code.

Yes, to mention that if **they provide custom OkHttpClient** - then they need to set all OkHttpClient-related options directly on OkHttpClient instance instead of passing it to `options.connection`.


> This is fine for me, but would be a breaking change for other users. Is that ok with you?

This library is more less is in an early stage. I think this API change wouldn't be big. Also, we could make small migration guide in release notes.

@neworld , what do you think about all of this?

neworld replied:
I like this idea as it adds more transparency.

chrisfillmore replied:
Ok I will update with these changes!

chrisfillmore replied:
Updated now... please review closely to be sure I haven't removed anything I shouldn't remove, and made the correct changes. Thanks.

chrisfillmore committed:
Remove Connection.Options.okHttpClientFactory

Previously, users could pass a lambda which supplied an OkHttpClient
instance. That has been removed, and replaced instead with an option
for users to pass their own OkHttp WebSocket.Factory.

README has also been updated to inform users how to use the updated
API.

GediminasZukas commented on library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt:
```diff
@@ -17,30 +13,24 @@ class Connection constructor(
      */
     data class Options(
-            var sslContext: SSLContext? = null,
-            var hostnameVerifier: HostnameVerifier? = null,
```
Actually we could keep those parameters

neworld replied:
@GediminasZukas, what is the point to keep those? In my opinion, they are adding the same confusion as `okHttpClientFactory`.

neworld replied:
Also, we could make a simple API for newcomers by providing some simple factory, which could be used as default method:
```
class SimpleFactory(val sslContext, val hostnameVerifier): WebSocket.Factory {
  ...
}
```

GediminasZukas replied:
> @GediminasZukas, what is the point to keep those?

I don't know exact uses cases with those. But can be that people use it. It is a legacy from original library. I would suggest to not touch those params until implementation of: https://github.com/vinted/actioncable-client-kotlin/issues/23

chrisfillmore replied:
I think it makes sense to put them back, because it will reduce the amount of breaking changes in this PR. More changes can come in a future update.

@GediminasZukas @neworld what do you guys want to do?

neworld replied:
Make sense. Let's keep it.

GediminasZukas replied:
@chrisfillmore , let's keep those parameters. Would be nice to implement this for now: https://github.com/vinted/actioncable-client-kotlin/pull/22#discussion_r559962495

chrisfillmore replied:
Ok I have updated.

GediminasZukas commented on library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Connection.kt:
```diff
@@ -139,12 +123,9 @@ class Connection constructor(
 
-        val httpClient = httpClientBuilder.build()
-        val webSocketFactory = options.okHttpWebSocketFactory ?: httpClient
+        val webSocketFactory = options.webSocketFactory ?: OkHttpClient()
```
... and apply it on fallback implementation:

```
?: OkHttpClient()
.newBuilder()
.sslSocketFactory(options.sslContext.socketFactory)
.hostnameVerifier(options.hostnameVerifier)
.build()
```
And inside documentation we could add explanation:

> Note: if you provide custom `WebSocket.Factory` implementation, then `sslContext` and `hostnameVerifier` options are ignored. You must set those options directly in your custom implementation instead of setting it on `options.connection`.

GediminasZukas reviewed (approved):
@chrisfillmore , thanks again! I will merge it and release library update soon.

chrisfillmore commented:
Thank you both for the review!
